tools/pyi/gen_pyi.py

```diff
@@ -0,0 +1,249 @@
+from __future__ import print_function
+import os
+import sys
+import inspect
+
+needed_modules = set()
+
+
+def type_to_python(typename):
+    typename = {'int64_t': 'int',
+                'Scalar': 'Union[float, int]',
+                'ScalarType': 'dtype',
+                'Device': 'Union[device, str, None]',
+                'IntList': 'Union[Tuple[int, ...], List[int, ...]]',
+                'Tensor?': 'Optional[Tensor]',
+                'Tensor': 'Tensor',
+                'bool': 'bool',
+                'double': 'float',
+                'Generator *': 'Generator',
+                'Generator*': 'Generator',
+                'std::vector<Tensor>': 'Union[Tuple[Tensor, ...],List[Tensor, ...]]',
+                'TensorList': 'Union[Tuple[Tensor, ...],List[Tensor, ...]]',
+                'Storage': 'Storage',  # inaccurate (FloatStorage...)
+                'SparseTensorRef': 'Tensor',
+                'void': 'None',
+                'Layout': 'layout',
+                'void*': 'int',  # dataptr
+                'std::string': 'str',
+                'real': 'Union[float, int]',
+                'accreal': 'Union[float, int]',
+                'IntegerTensor': 'Tensor',
+                'BoolTensor': 'Tensor',
+                'IndexTensor': 'Tensor',
+                }[typename]
+    # handle IntList[]
+    return typename
+
+
+def arg_to_type_hint(arg):
+    name = arg['name']
+    if name == 'from':  # keyword...
+        name += '_'
+    typename = type_to_python(arg['dynamic_type'])
+    if arg.get('is_nullable'):
+        typename = 'Optional[' + typename + ']'
+    if 'default' in arg:
+        default = '=' + str(arg['default'])
+    else:
+        default = ''
+    return name + ': ' + typename + default
+
+
+def generate_type_hints(fname, decls, is_tensor=False):
```

Comment:
So, is it right to say that this applies exclusively to methods, or are functions also annotated too?

Reply:
Do you mean as presented to the user or as technically implemented/as the type hints suggest?
It is called for both torch.XXX (is_tensor=False) and torch.Tensor.XXX. What happens below is that the functions torch.XXX are technically static methods in _C._VariableFunctions.